pkg/reporting/report.go

```diff
@@ -0,0 +1,251 @@
+	return ReportSecretKey
+}
+
+func (p *PreflightReportEvent) GetReportEventLimit() int {
+	return ReportEventLimit
+}
+
+func (p *PreflightReportEvent) GetReportMtx() *sync.Mutex {
+	return &preflightReportMtx
+}
+
+func CreateReportEvent(clientset kubernetes.Interface, namespace string, appSlug string, event ReportEvent) error {
+	event.GetReportMtx().Lock()
+	defer event.GetReportMtx().Unlock()
+
+	existingSecret, err := clientset.CoreV1().Secrets(namespace).Get(context.TODO(), event.GetReportSecretName(appSlug), metav1.GetOptions{})
+	if err != nil && !kuberneteserrors.IsNotFound(err) {
+		return errors.Wrap(err, "failed to get report secret")
+	} else if kuberneteserrors.IsNotFound(err) {
+		report := &Report{
+			Events: []ReportEvent{event},
+		}
+		data, err := EncodeReport(report)
+		if err != nil {
+			return errors.Wrap(err, "failed to encode report")
+		}
+
+		secret := &corev1.Secret{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "v1",
+				Kind:       "Secret",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      event.GetReportSecretName(appSlug),
+				Namespace: namespace,
+				Labels:    kotsadmtypes.GetKotsadmLabels(),
+			},
+			Data: map[string][]byte{
+				event.GetReportSecretKey(): data,
+			},
+		}
+
+		_, err = clientset.CoreV1().Secrets(namespace).Create(context.TODO(), secret, metav1.CreateOptions{})
+		if err != nil {
+			return errors.Wrap(err, "failed to create report secret")
+		}
+
+		return nil
+	}
+
+	if existingSecret.Data == nil {
+		existingSecret.Data = map[string][]byte{}
+	}
+
+	existingReport := &Report{}
+	if existingSecret.Data[event.GetReportSecretKey()] != nil {
+		if err := DecodeReport(existingSecret.Data[event.GetReportSecretKey()], existingReport, event.GetReportType()); err != nil {
+			return errors.Wrap(err, "failed to load existing report")
+		}
+	}
+
+	existingReport.Events = append(existingReport.Events, event)
+	if len(existingReport.Events) > event.GetReportEventLimit() {
+		existingReport.Events = existingReport.Events[len(existingReport.Events)-event.GetReportEventLimit():]
+	}
+
+	data, err := EncodeReport(existingReport)
+	if err != nil {
+		return errors.Wrap(err, "failed to encode existing report")
+	}
+
+	existingSecret.Data[event.GetReportSecretKey()] = data
+
+	_, err = clientset.CoreV1().Secrets(namespace).Update(context.TODO(), existingSecret, metav1.UpdateOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to update instance report secret")
+	}
+
+	return nil
+}
+
+func EncodeReport(r *Report) ([]byte, error) {
+	data, err := json.Marshal(r)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal report")
+	}
+	compressedData, err := util.GzipData(data)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to gzip report")
+	}
+	encodedData := base64.StdEncoding.EncodeToString(compressedData)
+
+	return []byte(encodedData), nil
+}
+
+func DecodeReport(encodedData []byte, existingReport *Report, reportType string) error {
```

Comment:
nit pick: i would suggest having this function return a decoded report instead of modifying `existingReport`.